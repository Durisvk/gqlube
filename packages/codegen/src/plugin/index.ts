import ts from "typescript";
import {
  PluginFunction,
  getCachedDocumentNodeFromSchema,
} from "@graphql-codegen/plugin-helpers";
import {
  GraphQLSchema,
  visit,
  Visitor,
  ObjectTypeDefinitionNode,
  EnumTypeDefinitionNode,
  UnionTypeDefinitionNode,
  InputObjectTypeDefinitionNode,
  ScalarTypeDefinitionNode,
} from "graphql";

const GraphQLScalars = {
  ID: ts.SyntaxKind.StringKeyword,
  String: ts.SyntaxKind.StringKeyword,
  Boolean: ts.SyntaxKind.BooleanKeyword,
  Int: ts.SyntaxKind.NumberKeyword,
  Float: ts.SyntaxKind.NumberKeyword,
};

export type Node =
  | ObjectTypeDefinitionNode
  | EnumTypeDefinitionNode
  | UnionTypeDefinitionNode
  | InputObjectTypeDefinitionNode
  | ScalarTypeDefinitionNode;

const log: any = [];

export const plugin: PluginFunction<any, any> = (schema: GraphQLSchema) => {
  const astNode = getCachedDocumentNodeFromSchema(schema);

  const typeDefs: {
    [key: string]: Node;
  } = {};

  const visitor: Visitor<{
    [key: string]: Node;
  }> = {
    leave: {
      ObjectTypeDefinition(node) {
        typeDefs[node.name.value] = node;
      },
      EnumTypeDefinition(node) {
        typeDefs[node.name.value] = node;
      },
      UnionTypeDefinition(node) {
        typeDefs[node.name.value] = node;
      },
      InputObjectTypeDefinition(node) {
        typeDefs[node.name.value] = node;
      },
      ScalarTypeDefinition(node) {
        typeDefs[node.name.value] = node;
      },
    },
  };

  visit(astNode, visitor as unknown as any);

  const nodes: (ts.TypeAliasDeclaration | ts.EnumDeclaration)[] = [];
  const printer = ts.createPrinter();

  for (const typeName of Object.keys(typeDefs)) {
    const typeDef = typeDefs[typeName];
    const result = createTypeScriptDefinitionFromGraphqlNode(typeDef);
    if (result) nodes.push(result);
  }

  log.push(typeDefs);

  return `${printer.printList(
    ts.ListFormat.MultiLine,
    ts.factory.createNodeArray(nodes),
    ts.createSourceFile("", "", ts.ScriptTarget.Latest)
  )}\n${JSON.stringify(log, null, 2)}`;
};

export function createTypeScriptDefinitionFromGraphqlNode(node: Node) {
  if (node.kind === "ObjectTypeDefinition") {
    return createTypeScriptDefinitionFromObjectType(node);
  } else if (node.kind === "EnumTypeDefinition") {
    return createTypeScriptDefinitionFromEnumType(node);
  } else if (node.kind === "UnionTypeDefinition") {
    return createTypeScriptDefinitionFromUnionType(node);
  } else if (node.kind === "InputObjectTypeDefinition") {
    return createTypeScriptDefinitionFromInputObjectType(node);
  } else if (node.kind === "ScalarTypeDefinition") {
    return createTypeScriptDefinitionFromScalarType(node);
  }
}

export function createMembers(
  fieldNames: string[],
  fieldTypes: any[],
  fieldArguments: any[] = []
) {
  return fieldNames.map((fieldName, index) => {
    const fieldType = fieldTypes[index];
    const type = fieldType.kind === "NonNullType" ? fieldType.type : fieldType;
    const isList = type.kind === "ListType";
    const isNonNull = fieldType.kind === "NonNullType";
    const tsType = isList ? createListType(type.type) : createScalarType(type);
    const args = fieldArguments[index];

    const tsProperty = ts.factory.createPropertySignature(
      undefined,
      fieldName,
      isNonNull
        ? undefined
        : ts.factory.createToken(ts.SyntaxKind.QuestionToken),
      tsType
    );

    if (args && args.length) {
      const functionType = createFunctionType(args, tsType);
      return ts.factory.createPropertySignature(
        undefined,
        fieldName,
        undefined,
        functionType
      );
    }

    return tsProperty;
  });
}

export function createListType(type: any): ts.TypeNode {
  if (type.kind === "NamedType") {
    const name = type.name.value;
    if (name in GraphQLScalars) {
      return ts.factory.createArrayTypeNode(
        ts.factory.createKeywordTypeNode(
          GraphQLScalars[
            name as keyof typeof GraphQLScalars
          ] as ts.KeywordTypeSyntaxKind
        )
      );
    } else {
      return ts.factory.createArrayTypeNode(
        ts.factory.createTypeReferenceNode(name, undefined)
      );
    }
  }

  if (type.kind === "ListType") {
    return ts.factory.createArrayTypeNode(createListType(type.type));
  }

  if (type.kind === "NonNullType") {
    return createListType(type.type);
  }

  return ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword);
}

export function createScalarType(type: any) {
  if (type.kind === "NamedType") {
    const name = type.name.value;
    if (name in GraphQLScalars) {
      return ts.factory.createKeywordTypeNode(
        GraphQLScalars[
          name as keyof typeof GraphQLScalars
        ] as ts.KeywordTypeSyntaxKind
      );
    } else {
      return ts.factory.createTypeReferenceNode(name, undefined);
    }
  }
}

export function createFunctionType(args: any[], type?: ts.TypeNode) {
  return ts.factory.createFunctionTypeNode(
    undefined,
    [
      ts.factory.createParameterDeclaration(
        undefined,
        undefined,
        "inputs",
        args.some((arg) => arg.type.kind === "NonNullType")
          ? undefined
          : ts.factory.createToken(ts.SyntaxKind.QuestionToken),
        ts.factory.createTypeLiteralNode(
          args.map((arg) => {
            const isNonNull = arg.type.kind === "NonNullType";
            const argType = isNonNull ? arg.type.type : arg.type;
            const isList = argType.kind === "ListType";
            const tsType = isList
              ? createListType(argType.type)
              : createScalarType(argType);

            const graphqlType = getGraphQLType(argType);

            return ts.factory.createPropertySignature(
              undefined,
              `'${arg.name.value}: ${graphqlType}'`,
              isNonNull
                ? undefined
                : ts.factory.createToken(ts.SyntaxKind.QuestionToken),
              tsType
            );
          })
        )
      ),
    ],
    type || ts.factory.createKeywordTypeNode(ts.SyntaxKind.VoidKeyword)
  );
}

export function createTypeScriptDefinitionFromObjectType(
  node: ObjectTypeDefinitionNode
) {
  const fields = node?.fields;
  const fieldNames = fields?.map((field) => field.name.value) || [];
  const fieldTypes = fields?.map((field) => field.type) || [];
  const fieldArguments = fields?.map((field) => field.arguments) || [];

  const members = createMembers(fieldNames, fieldTypes, fieldArguments);
  const tsType = ts.factory.createTypeLiteralNode([
    ts.factory.createPropertySignature(
      undefined,
      "__typename",
      undefined,
      ts.factory.createLiteralTypeNode(
        ts.factory.createStringLiteral(node.name.value)
      )
    ),
    ...members,
  ]);

  return ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    node.name.value,
    undefined,
    tsType
  );
}

export function createTypeScriptDefinitionFromEnumType(
  node: EnumTypeDefinitionNode
) {
  const members =
    node.values?.map((value) =>
      ts.factory.createEnumMember(value.name.value)
    ) || [];

  const type = ts.factory.createEnumDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    node.name.value,
    members
  );
  return type;
}

export function createTypeScriptDefinitionFromUnionType(
  node: UnionTypeDefinitionNode
) {
  const members =
    node.types?.map((type) =>
      ts.factory.createTypeReferenceNode(type.name.value, undefined)
    ) || [];

  const type = ts.factory.createUnionTypeNode(members);
  return ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    node.name.value,
    undefined,
    type
  );
}

export function createTypeScriptDefinitionFromInputObjectType(
  node: InputObjectTypeDefinitionNode
) {
  const fields = node?.fields;
  const fieldNames = fields?.map((field) => field.name.value) || [];
  const fieldTypes = fields?.map((field) => field.type) || [];
  // handle non null types, list types, scalars and type aliases
  const members = createMembers(fieldNames, fieldTypes);
  const type = ts.factory.createTypeLiteralNode(members);
  return ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    node.name.value,
    undefined,
    type
  );
}

function createTypeScriptDefinitionFromScalarType(
  node: ScalarTypeDefinitionNode
) {
  const type = ts.factory.createTypeAliasDeclaration(
    [ts.factory.createModifier(ts.SyntaxKind.ExportKeyword)],
    node.name.value,
    undefined,
    ts.factory.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword)
  );
  return type;
}

export function getGraphQLType(type: any): string | undefined {
  if (type.kind === "NamedType") {
    return type.name.value;
  }

  if (type.kind === "ListType") {
    return `[${getGraphQLType(type.type)}]`;
  }

  if (type.kind === "NonNullType") {
    return `${getGraphQLType(type.type)}!`;
  }
}
