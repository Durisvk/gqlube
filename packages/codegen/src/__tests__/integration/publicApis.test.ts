import ts from "typescript";
import { createTypeScriptDefinitionFromGraphqlNode, Node } from "../../plugin";
import rickandmortyApi from "./rickandmortyapi.json";
import countriesApi from "./countriesapi.json";

describe("Public APIs", () => {
  const processApi = (apiTypeDefs: any) => {
    const nodes: Node[] = [];
    for (const typeName of Object.keys(apiTypeDefs)) {
      const typeDef = apiTypeDefs[typeName as keyof typeof apiTypeDefs];
      const result = createTypeScriptDefinitionFromGraphqlNode(typeDef);
      if (result) nodes.push(result as any);
    }
    const printer = ts.createPrinter();
    const resultingTypescript = printer.printList(
      ts.ListFormat.MultiLine,
      ts.factory.createNodeArray(nodes as any),
      ts.createSourceFile("", "", ts.ScriptTarget.Latest)
    );
    return resultingTypescript;
  };

  it("https://rickandmortyapi.com/graphql (rickandmortyapi.json) should be processable by the plugin", () => {
    expect(processApi(rickandmortyApi)).toMatchSnapshot();
  });

  it("https://countries.trevorblades.com (countriesapi.json) should be processable by the plugin", () => {
    expect(processApi(countriesApi)).toMatchSnapshot();
  });
});
