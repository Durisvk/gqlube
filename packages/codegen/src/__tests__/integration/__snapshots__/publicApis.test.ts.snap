// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Public APIs https://countries.trevorblades.com (countriesapi.json) should be processable by the plugin 1`] = `
"export type Query = {
    __typename: \\"Query\\";
    _entities: (inputs: {
        'representations: [_Any!]': _Any[];
    }) => _Entity[];
    _service: _Service;
    continent: (inputs: {
        'code: ID': string;
    }) => Continent;
    continents: (inputs?: {
        'filter: ContinentFilterInput'?: ContinentFilterInput;
    }) => Continent[];
    countries: (inputs?: {
        'filter: CountryFilterInput'?: CountryFilterInput;
    }) => Country[];
    country: (inputs: {
        'code: ID': string;
    }) => Country;
    language: (inputs: {
        'code: ID': string;
    }) => Language;
    languages: (inputs?: {
        'filter: LanguageFilterInput'?: LanguageFilterInput;
    }) => Language[];
};
export type _Any = any;
export type _Entity = Continent | Country | Language;
export type Continent = {
    __typename: \\"Continent\\";
    code: string;
    countries: Country[];
    name: string;
};
export type Country = {
    __typename: \\"Country\\";
    capital?: string;
    code: string;
    continent: Continent;
    currency?: string;
    emoji: string;
    emojiU: string;
    languages: Language[];
    name: string;
    native: string;
    phone: string;
    states: State[];
};
export type Language = {
    __typename: \\"Language\\";
    code: string;
    name?: string;
    native?: string;
    rtl: boolean;
};
export type State = {
    __typename: \\"State\\";
    code?: string;
    country: Country;
    name: string;
};
export type _Service = {
    __typename: \\"_Service\\";
    sdl?: string;
};
export type ContinentFilterInput = {
    code?: StringQueryOperatorInput;
};
export type StringQueryOperatorInput = {
    eq?: string;
    glob?: string;
    in?: string[];
    ne?: string;
    nin?: string[];
    regex?: string;
};
export type CountryFilterInput = {
    code?: StringQueryOperatorInput;
    continent?: StringQueryOperatorInput;
    currency?: StringQueryOperatorInput;
};
export type LanguageFilterInput = {
    code?: StringQueryOperatorInput;
};
"
`;

exports[`Public APIs https://rickandmortyapi.com/graphql (rickandmortyapi.json) should be processable by the plugin 1`] = `
"export type Query = {
    __typename: \\"Query\\";
    character: (inputs: {
        'id: ID': string;
    }) => Character;
    characters: (inputs?: {
        'filter: FilterCharacter'?: FilterCharacter;
        'page: Int'?: number;
    }) => Characters;
    charactersByIds: (inputs: {
        'ids: [ID!]': string[];
    }) => Character[];
    episode: (inputs: {
        'id: ID': string;
    }) => Episode;
    episodes: (inputs?: {
        'filter: FilterEpisode'?: FilterEpisode;
        'page: Int'?: number;
    }) => Episodes;
    episodesByIds: (inputs: {
        'ids: [ID!]': string[];
    }) => Episode[];
    location: (inputs: {
        'id: ID': string;
    }) => Location;
    locations: (inputs?: {
        'filter: FilterLocation'?: FilterLocation;
        'page: Int'?: number;
    }) => Locations;
    locationsByIds: (inputs: {
        'ids: [ID!]': string[];
    }) => Location[];
};
export type Character = {
    __typename: \\"Character\\";
    created?: string;
    episode: Episode[];
    gender?: string;
    id?: string;
    image?: string;
    location?: Location;
    name?: string;
    origin?: Location;
    species?: string;
    status?: string;
    type?: string;
};
export type Episode = {
    __typename: \\"Episode\\";
    air_date?: string;
    characters: Character[];
    created?: string;
    episode?: string;
    id?: string;
    name?: string;
};
export type Location = {
    __typename: \\"Location\\";
    created?: string;
    dimension?: string;
    id?: string;
    name?: string;
    residents: Character[];
    type?: string;
};
export type FilterCharacter = {
    gender?: string;
    name?: string;
    species?: string;
    status?: string;
    type?: string;
};
export type Characters = {
    __typename: \\"Characters\\";
    info?: Info;
    results?: Character[];
};
export type Info = {
    __typename: \\"Info\\";
    count?: number;
    next?: number;
    pages?: number;
    prev?: number;
};
export type FilterEpisode = {
    episode?: string;
    name?: string;
};
export type Episodes = {
    __typename: \\"Episodes\\";
    info?: Info;
    results?: Episode[];
};
export type FilterLocation = {
    dimension?: string;
    name?: string;
    type?: string;
};
export type Locations = {
    __typename: \\"Locations\\";
    info?: Info;
    results?: Location[];
};
export enum CacheControlScope {
    PRIVATE,
    PUBLIC
}
"
`;
